server/chat-plugins/youtube.ts

```diff
@@ -0,0 +1,422 @@
+} catch (e) {
+	channelData = {};
+}
+
+export const YouTube = new class {
+	interval: NodeJS.Timeout | null;
+	constructor() {
+		this.interval = null;
+	}
+	async getChannelData(channel: string, username?: string) {
+		const queryUrl = `${CHANNEL}?part=snippet%2Cstatistics&id=${encodeURIComponent(channel)}&key=${Config.youtubeKey}`;
+		const query = new Promise((resolve, reject) => {
+			https.get(queryUrl, res => {
+				const data: string[] = [];
+				res.setEncoding('utf8');
+				res.on('data', chunk => data.push(chunk));
+				res.on('end', () => resolve(JSON.parse(data.join(''))));
+			}).on('error', reject);
+		});
+		const res: any = await query.catch(() => {});
+		if (!res) return null;
+		const data = res.items[0];
+		const cache = {
+			name: data.snippet.title,
+			description: data.snippet.description,
+			url: data.snippet.customUrl,
+			icon: data.snippet.thumbnails.medium.url,
+			videos: Number(data.statistics.videoCount),
+			subs: Number(data.statistics.subscriberCount),
+			views: Number(data.statistics.viewCount),
+			username: username,
+		};
+		channelData[channel] = {...cache};
+		FS(STORAGE_PATH).writeUpdate(() => JSON.stringify(channelData));
+		return cache;
+	}
+	async generateChannelDisplay(id: string) {
+		// url isn't needed but it destructures wrong without it
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars
+		const {name, description, url, icon, videos, subs, views, username} = await this.get(id);
+		// credits asheviere for most of the html
+		let buf = `<div class="infobox"><table style="margin:0px;"><tr>`;
+		buf += `<td style="margin:5px;padding:5px;min-width:175px;max-width:160px;text-align:center;border-bottom:0px;">`;
+		buf += `<div style="padding:5px;background:white;border:1px solid black;margin:auto;max-width:100px;max-height:100px;">`;
+		buf += `<a href="${ROOT}channel/${id}"><img src="${icon}" width=100px height=100px/></a>`;
+		buf += `</div><p style="margin:5px 0px 4px 0px;word-wrap:break-word;">`;
+		buf += `<a style="font-weight:bold;color:#c70000;font-size:12pt;" href="https://www.youtube.com/channel/${id}">${name}</a>`;
+		buf += `</p></td><td style="padding: 0px 25px;font-size:10pt;background:rgb(220,20,60);width:100%;border-bottom:0px;vertical-align:top;">`;
+		buf += `<p style="padding: 5px;border-radius:8px;color:white;font-weight:bold;text-align:center;">`;
+		buf += `${videos} videos | ${subs} subscribers | ${views} video views</p>`;
+		buf += `<p style="margin-left: 5px; font-size:9pt;color:white;">`;
+		buf += `${description.slice(0, 400).replace(/\n/g, ' ')}${description.length > 400 ? '(...)' : ''}</p>`;
+		if (username) {
+			buf += `<p style="text-align:left;font-style:italic;color:white;">PS username: ${username}</p></td></tr></table></div>`;
+		} else {
+			buf += '</td></tr></table></div>';
+		}
+		return buf;
+	}
+	randChannel() {
+		const keys = Object.keys(channelData);
+		const id = keys[Math.floor(Math.random() * keys.length)].trim();
```

Comment:
```suggestion
		const id = Dex.shuffle(keys)[0].trim();
```

(this is optional, i just think that if we have a function, why not use it. though then it brings the question of why `sample` is on the `PRNG` class and not on `Dex` as well)